File: Sources/ThesisSkipApp/data/repository/UserRepository.swift
//
//  UserRepository.swift
//  MADiOS
//
//  Created by Bram Rampelberg on 12/08/2025.
//  Copyright © 2025 HOGENT. All rights reserved.
//

import Foundation

final class UserRepository {
    nonisolated(unsafe) static let shared = UserRepository()

    private let userService = UserService.shared

    func getUser() async -> Result<User> {
        let result = await userService.fetchUserDetails()
        if result.isSuccess {
            return .success(resultData: User(fromDto: result.data!))
        } else {
            #if SKIP
                return failureWithLog(
                    cause: "Failed to get user data.",
                    error: result.error
                )
            #else
                return .failureWithLog(
                    cause: "Failed to get user data.",
                    error: result.error
                )
            #endif
        }

    }
}

File: .build/plugins/outputs/thesis-skip-app/ThesisSkipApp/destination/skipstone/ThesisSkipApp/src/main/kotlin/thesis/skip/app/UserRepository.kt
//
//  UserRepository.swift
//  MADiOS
//
//  Created by Bram Rampelberg on 12/08/2025.
//  Copyright © 2025 HOGENT. All rights reserved.
//

package thesis.skip.app

import skip.lib.*

import skip.foundation.*

internal class UserRepository {

    private val userService = UserService.shared

    internal suspend fun getUser(): Result<User> = Async.run l@{
        val result = userService.fetchUserDetails()
        if (result.isSuccess) {
            return@l Result.success(resultData = User(fromDto = result.data!!))
        } else {
            return@l failureWithLog(cause = "Failed to get user data.", error = result.error)
        }

    }

    companion object {
        internal val shared = UserRepository()
    }
}
